<?xml version="1.0"?>
<launch>
  <arg name="control_mode" default="3"/>
  <arg name="debug_motor" default="false"/>
  <arg name="debug_imu" default="false"/>
  <arg name="publish_TF" default="false"/>
  <arg name="publish_odom" default="true"/>
  <arg name="use_sim_time" default="false"/>
  <arg name="twist_cmd_vel_topic" default="/cmd_vel"/>
  <arg name="cmd_vel_topic" default="/vel/cmd_vel"/>
  <arg name="cmd_rpm_topic" default="/vel/cmd_rpm"/>
  <arg name="cmd_deg_topic" default="/pos/cmd_deg"/>
  <arg name="cmd_dist_topic" default="/pos/cmd_dist"/>
  <arg name="xacro_file_path" default="$(find-pkg-share orange_description)/xacro/orange_robot.xacro"/>
  <arg name="motor_driver_config_file_path" default="$(find-pkg-share orange_bringup)/config/motor_driver_params.yaml"/>

  <!-- robot_state_publisher -->
  <node pkg="robot_state_publisher" exec="robot_state_publisher">
    <param name="robot_description" value="$(command 'xacro $(var xacro_file_path)')"/>
  </node>
  <!-- joint_state_publisher -->
  <node pkg="joint_state_publisher" exec="joint_state_publisher">
    <param name="robot_description" value="$(command 'xacro $(var xacro_file_path)')"/>
  </node>

  <!-- motor_driver_node -->
  <node pkg="orange_bringup" exec="motor_driver_node" output="screen">
    <param from="$(var motor_driver_config_file_path)"/>
    <param name="port" value="/dev/ZLAC8015D"/>
    <param name="control_mode" value="$(var control_mode)"/>
    <param name="debug" value="$(var debug_motor)"/>
    <param name="twist_cmd_vel_topic" value="$(var twist_cmd_vel_topic)"/>
    <param name="cmd_vel_topic" value="$(var cmd_vel_topic)"/>
    <param name="cmd_rpm_topic" value="$(var cmd_rpm_topic)"/>
    <param name="cmd_deg_topic" value="$(var cmd_deg_topic)"/>
    <param name="cmd_dist_topic" value="$(var cmd_dist_topic)"/>
    <param name="publish_TF" value="$(var publish_TF)"/>
    <param name="TF_header_frame" value="odom"/>
    <param name="TF_child_frame" value="base_footprint"/>
    <param name="publish_odom" value="$(var publish_odom)"/>
    <param name="odom_header_frame" value="odom"/>
    <param name="odom_child_frame" value="base_footprint"/>
  </node>
  <!-- estop  
  <node pkg="estop_ros" exec="cmd_vel_override_node" output="screen">
    <param name="port" value="/dev/sensors/estop"/>
    <param name="baudrate" value="115200"/>
    <param name="time_out" value="500"/>
    <param name="serial_interval" value="0.01"/>
    <remap from="/estop/state" to="/estop"/>
  </node>-->

  <!-- imu -->
  <include file="$(find-pkg-share icm_20948)/launch/run.launch.xml">
    <arg name="port" value="/dev/sensors/imu"/>
    <arg name="time_out" value="0.5"/>
    <arg name="baudrate" value="115200"/>
    <arg name="imu_topic" value="imu"/>
    <arg name="frame_id" value="imu_link"/>
    <arg name="debug" value="$(var debug_imu)"/>
  </include>

  <!-- spresense imu -->
  <node pkg="cxd5602pwbimu" exec="spresense_imu_node">
    <param name="port" value="/dev/sensors/spresense_imu"/>
    <param name="frame_id" value="imu_link"/>
    <param name="imu_topic" value="/imu/spresense"/>
    <param name="time_out" value="10.0"/>
    <param name="baudrate" value="921600"/>
  </node>

  <!-- mid360 -->
  <include file="$(find-pkg-share livox_ros_driver2)/launch_ROS2/msg_MID360_launch.py"/>
</launch>
